File: backend/reasoning/causal_graph.py
"""
Darukaa.Earth AI Biodiversity Intelligence System
Causal Graph & Mechanistic Explanation Generator
Builds verifiable causal chains linking physical interventions to biochemical and systemic ecological outcomes.
"""

from typing import List, Dict, Any

class CausalChain:
    def __init__(self, step_title: str, pathway: str, impacted_variable: str, biochemical_evidence: str):
        self.step_title = step_title
        self.pathway = pathway
        self.impacted_variable = impacted_variable
        self.biochemical_evidence = biochemical_evidence

    def to_dict(self) -> Dict[str, str]:
        return {
            "step_title": self.step_title,
            "pathway": self.pathway,
            "impacted_variable": self.impacted_variable,
            "biochemical_evidence": self.biochemical_evidence
        }

def generate_causal_chains(scenario_type: str) -> List[Dict[str, Any]]:
    """
    Produces deterministic, scientifically grounded causal chains for ecological interventions.
    """
    if scenario_type == "semi_arid_agroforestry_wheat":
        return [
            {
                "chain_id": "CHAIN-01",
                "name": "Soil Organic Carbon ↔ Moisture Retention ↔ Microbial Rhizosphere",
                "nodes": [
                    {
                        "stage": "1. Input / Action",
                        "title": "Reverse-Phenology Agroforestry & Leguminous Intercropping",
                        "description": "Establish Faidherbia albida (30 trees/ha) + Cajanus cajan (pigeon pea) interrow planting in wheat monoculture."
                    },
                    {
                        "stage": "2. Biochemical Transformation",
                        "title": "Biological N2 Fixation & Glomalin Exudation",
                        "description": "Rhizobia nodules synthesize bioavailable ammonium; mycorrhizal fungi secrete glomalin glycoprotein binding soil mineral grains into stable micro-aggregates."
                    },
                    {
                        "stage": "3. Hydrological & Structural Delta",
                        "title": "Pore Distribution & Available Water Capacity (+32%)",
                        "description": "Increases macroporosity and capillary retention, extending root zone moisture availability by 35-45 days under semi-arid rainfall regimes."
                    },
                    {
                        "stage": "4. Ecosystem Climax",
                        "title": "Microbiome Biomass (+55%) & Shannon Biodiversity ($H'$ +0.85)",
                        "description": "Diverse heterotrophic bacterial guilds and fungal hyphal networks buffer wheat crops against heat desiccation and nutrient starvation."
                    }
                ]
            },
            {
                "chain_id": "CHAIN-02",
                "name": "Microclimate Canopy Buffering ↔ Evapotranspiration ↔ Yield Stability",
                "nodes": [
                    {
                        "stage": "1. Input / Action",
                        "title": "Multi-Strata Agroforestry Canopy Structuring",
                        "description": "Layered vegetation canopy creating partial diffuse shading and aerodynamic wind resistance."
                    },
                    {
                        "stage": "2. Thermodynamic Response",
                        "title": "Vapor Pressure Deficit (VPD) Reduction & Thermal Moderation",
                        "description": "Transpirational cooling lowers topsoil and ambient canopy temperatures by 2.2–3.8°C during peak mid-day solar radiation."
                    },
                    {
                        "stage": "3. Systemic Resilience",
                        "title": "Wheat Stomatal Conductance Maintenance & Yield Stabilization (+28%)",
                        "description": "Prevents photosynthetic shutdown during acute dryland thermal pulses, securing multi-year harvest stability (IPCC AR6 WGII)."
                    }
                ]
            }
        ]
    elif scenario_type == "acidic_tropical_pasture":
        return [
            {
                "chain_id": "CHAIN-01",
                "name": "Biochar Co-Composting ↔ Cation Exchange ↔ Root Colonization",
                "nodes": [
                    {
                        "stage": "1. Input / Action",
                        "title": "Inoculated Woody Biochar (5 t/ha) + Multi-species Legume Forage",
                        "description": "Broadcast pyrolyzed carbon inoculated with mycorrhizal fungi and humified compost."
                    },
                    {
                        "stage": "2. Chemical Buffering",
                        "title": "Al3+ Toxicity Immobilization & Base Saturation Increase",
                        "description": "Carboxylate functional groups buffer pH from 4.8 to 6.3, exchanging essential Ca2+/Mg2+ cations."
                    },
                    {
                        "stage": "3. Biodiversity Response",
                        "title": "Arbuscular Mycorrhizal Fungi (AMF) Colonization Surge (+55%)",
                        "description": "Hyphal networks bridge soil particles, increasing active organic matter and earthworm biomass."
                    }
                ]
            }
        ]
    else:
        # Generic multi-variable coupled causal chain
        return [
            {
                "chain_id": "CHAIN-01",
                "name": "Vegetative Stratification ↔ Soil Carbon ↔ Trophic Food Web",
                "nodes": [
                    {
                        "stage": "1. Action",
                        "title": "Diversified Polyculture & Native Perennial Margins",
                        "description": "Replace bare boundaries and continuous monoculture with multi-species functional groups."
                    },
                    {
                        "stage": "2. Biological Cascade",
                        "title": "Rhizodeposition & Continuous Nectar Secretion",
                        "description": "Root exudates nourish bacterial/fungal decomposers while floral resources sustain hoverflies, solitary bees, and parasitoid wasps."
                    },
                    {
                        "stage": "3. Ecosystem Metric Uplift",
                        "title": "Soil Organic Carbon (+22%) & Wild Pollinator Richness (+65%)",
                        "description": "Concurrently elevates soil moisture retention, reduces pesticide requirements by 70%, and stabilizes ecological network redundancy."
                    }
                ]
            }
        ]

File: backend/src/reasoning/causalGraph.js
/**
 * Darukaa.Earth AI Biodiversity Intelligence System
 * Causal Graph & Biological Mechanistic Pathway Engine (ES Modules)
 */

export function generateCausalChains(scenarioType) {
  if (scenarioType === "semi_arid_agroforestry_wheat") {
    return [
      {
        chain_id: "CHAIN-01",
        name: "Soil Organic Carbon ↔ Soil Porosity ↔ Rhizosphere Biome ↔ Water Retention",
        nodes: [
          {
            stage: "1. Prescribed Intervention",
            title: "Reverse-Phenology Agroforestry & Leguminous Intercropping",
            description: "Establish Faidherbia albida (30 trees/ha) + Cajanus cajan (pigeon pea) inter-row planting in monoculture wheat acreage."
          },
          {
            stage: "2. Biochemical Transformation",
            title: "Biological N2 Fixation & Glomalin Secretion",
            description: "Rhizobia nodule symbiosis fixes 35–55 kg bioavailable N/ha; arbuscular mycorrhizal fungi (AMF) exude glomalin glycoprotein binding mineral particles into water-stable macro-aggregates."
          },
          {
            stage: "3. Hydrological Structural Response",
            title: "Capillary Pore Volume & Available Water Capacity (+32%)",
            description: "Enhances root-zone water storage by 28–38%, buffering crops against 40+ days of mid-season precipitation gaps in semi-arid drylands."
          },
          {
            stage: "4. Ecosystem Climax",
            title: "Microbial Biomass (+55%) & Shannon Biodiversity ($H'$ +0.85)",
            description: "Heterogeneous root exudates nourish diverse bacterial and fungal decomposers, suppressing soil-borne pathogens and stabilizing grain yields."
          }
        ]
      },
      {
        chain_id: "CHAIN-02",
        name: "Microclimate Canopy Buffering ↔ Vapor Pressure Deficit ↔ Photosynthetic Continuity",
        nodes: [
          {
            stage: "1. Prescribed Intervention",
            title: "Stratified Canopy Architecture (Evergreen Agriculture)",
            description: "Spaced perennial tree canopy providing filtered solar radiance and wind attenuation."
          },
          {
            stage: "2. Thermodynamic Response",
            title: "Vapor Pressure Deficit (VPD) Reduction & Transpirational Cooling",
            description: "Reduces peak mid-day canopy surface temperatures by 2.2–3.8°C and suppresses direct evaporative moisture loss from topsoil by 20–35% (IPCC AR6 WGII)."
          },
          {
            stage: "3. Climax Stabilization",
            title: "Crop Stomatal Conductance & Drought Resilience (+30%)",
            description: "Prevents midday stomatal closure and thermal shock in wheat, securing harvest viability during drought pulses."
          }
        ]
      }
    ];
  }

  return [
    {
      chain_id: "CHAIN-01",
      name: "Plant Diversity ↔ Underground Hyphal Networks ↔ Trophic Food Webs",
      nodes: [
        {
          stage: "1. Intervention",
          title: "Polyculture Diversification & Native Perennial Boundary Margins",
          description: "Transition continuous single-crop fields to 4-family polycultures with 8% flowering hedgerow corridors."
        },
        {
          stage: "2. Biological Cascade",
          title: "Multi-Tier Rhizodeposition & Continuous Nectar Provision",
          description: "Variable root depths exude diverse polysaccharides feeding mycorrhizae; nectar flora feeds hoverflies, parasitoid wasps, and wild bees."
        },
        {
          stage: "3. Systemic Metric Improvement",
          title: "Soil Carbon (+25%) & Wild Pollinator Richness (+65%)",
          description: "Simultaneously increases soil water capacity, drives natural biological pest suppression, and restores regional biodiversity index."
        }
      ]
    }
  ];
}
